crates/sound/src/vhu_sound.rs

```diff
@@ -50,25 +53,181 @@ impl VhostUserSoundThread {
         Ok(())
     }
 
-    fn handle_event(&self, device_event: u16, vrings: &[VringRwLock]) -> IoResult<bool> {
+    fn handle_event(&self, device_event: u16, vrings: &[VringRwLock]) -> result::Result<bool, io::Error> {
         let vring = &vrings[device_event as usize];
         let queue_idx = self.queue_indexes[device_event as usize];
+        debug!("handle event call queue: {}", queue_idx);
 
         match queue_idx {
-            CONTROL_QUEUE_IDX => self.process_control(vring),
-            EVENT_QUEUE_IDX => self.process_event(vring),
-            TX_QUEUE_IDX => self.process_tx(vring),
-            RX_QUEUE_IDX => self.process_rx(vring),
-            _ => Err(Error::HandleUnknownEvent.into()),
+            CONTROL_QUEUE_IDX => {
+                debug!("control queue: {}", CONTROL_QUEUE_IDX);
+                let vring = &vrings[0];
+                if self.event_idx {
+                    // vm-virtio's Queue implementation only checks avail_index
+                    // once, so to properly support EVENT_IDX we need to keep
+                    // calling process_request_queue() until it stops finding
+                    // new requests on the queue.
+                    loop {
+                        vring.disable_notification().unwrap();
+                        self.process_control(vring)?;
+                        if !vring.enable_notification().unwrap() {
+                            break;
+                        }
+                    }
+                } else {
+                    // Without EVENT_IDX, a single call is enough.
+                    self.process_control(vring)?;
+                }
+            }
+            EVENT_QUEUE_IDX => {
+                self.process_event(vring)?;
+            }
+            TX_QUEUE_IDX => {
+                self.process_tx(vring)?;
+            }
+            RX_QUEUE_IDX => {
+                self.process_rx(vring)?;
+            }
+            _ => {
+                return Err(Error::HandleUnknownEvent.into());
+            }
         }
+        Ok(false)
     }
 
-    fn process_control(&self, _vring: &VringRwLock) -> IoResult<bool> {
-        Ok(false)
+    /// Process the messages in the vring and dispatch replies
+    fn process_control(&self, vring: &VringRwLock) -> Result<bool, Error> {
+        let requests: Vec<SndDescriptorChain> = vring
+            .get_mut()
+            .get_queue_mut()
+            .iter(self.mem.as_ref().unwrap().memory())
+            .map_err(|_| Error::DescriptorNotFound)?
+            .collect();
+
+        debug!("Requests to process: {}", requests.len());
+        if requests.is_empty() {
+            debug!("yes, it's empty");
+            return Ok(true);
+        }
+        //iterate over each sound request
+        for desc_chain in requests {
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+            debug!("Sound request with n descriptors: {}", descriptors.len());
+            if descriptors.len() != 3 {
```

Comment:
Since the number of descriptors depends on the request, I think the number of descriptors should be checked when evaluating the different control queue messages. `VIRTIO_SND_R_PCM_SET_PARAMS`, for example, has two descriptors.

Reply:
yes, make sense